### ISSUE TITLE
Is there anything missing for a stable release?

### ISSUE THREAD

**Issue by stax76** [staxrip/staxrip#420]
Is there anything missing for a stable release?

I've updated mkvtoolnix via auto update.

If not then maybe I should just upload it and hope it's not buggy.

**Comment by Neltulz** [comment]
I will take a peek at the latest build...

**Comment by Neltulz** [comment]
Okay, so far I have noticed 2 things out of the ordinary.

1. --input-bit-depth is not changing from 8 to 10 when a native 10bit video is loaded into StaxRip.  If you are going to forcibly specify --input-bit-depth, it needs to match the input video.  It appears aomenc will self-correct but produce a warning, so it's not a huge problem.

![image](https://user-images.githubusercontent.com/7741742/103689808-69c2f480-4f59-11eb-928e-a7d1b946db03.png)



2. CQ-Level is still in the "AV1 Specific 2" section, when IMO It should be in the "Rate Control 1" Section directly below "Rate Mode" for fast access.

![2021-01-05_13-21-21](https://user-images.githubusercontent.com/7741742/103689548-1355b600-4f59-11eb-9be8-288730e93c20.png)
![2021-01-05_13-22-02](https://user-images.githubusercontent.com/7741742/103689551-1486e300-4f59-11eb-862e-71c9ed3acf56.png)

3. Forcibly specifying --threads: 32 is kind of silly I take it, considering the response I got from the AV1 discord community.

![image](https://user-images.githubusercontent.com/7741742/103690201-f8d00c80-4f59-11eb-8899-88edb04ff9f6.png)


On user recommends doing this to inquire about maximum thread count so that it can be set accordingly.  Below he recommends a more precise way of acquiring threads, and he recommends capping thread count at 32.  He also linked this URL: https://en.cppreference.com/w/cpp/thread/thread/hardware_concurrency

![image](https://user-images.githubusercontent.com/7741742/103690125-d9d17a80-4f59-11eb-8781-56a96b821427.png)

4. aomenc produced an error saying that ``-width`` and ``--height`` must be specified.  I am not sure why aomenc would fail without those being specified, but you might want to automatically set these values from the input width and height to avoid this error:

![2021-01-05_13-33-56](https://user-images.githubusercontent.com/7741742/103692019-c83da200-4f5c-11eb-8acb-beb253690b01.png)

5. I was unable to even encode a video because of an error related to 2pass.  I have no idea what these even means, and the AV1 community is confused about it too.

![2021-01-05_13-34-47](https://user-images.githubusercontent.com/7741742/103692096-e6a39d80-4f5c-11eb-86ca-9144e89fc5e8.png)

6.  Unsure why PSNR is being set to 0.  A member of the AV1 community was confused about this.  Also, PSNR should be set to 1 because that's what you have stated in the drop down menu.
![image](https://user-images.githubusercontent.com/7741742/103692220-16eb3c00-4f5d-11eb-998d-7a4009243204.png)

All in all, I would say, it would be premature to make a stable build until these issues are resolved, as I am completely unable to encode a video right now with my settings and input video.

Here is a link to the video I am using for testing so you can perform your own tests:
https://www.dropbox.com/s/mpgb6ht0ijz2qh3/10bit_MKV.7z?dl=0

**Comment by Dendraspis** [comment]
1. Yeah, an automatic setting could be nice, indeed. At the moment it's just a default or better *the first possible value*. 
@stax76   Do you have the ambition to implement that?  😉 

2. Already done, but not out, yet.  But this is an exception - at the moment we stick to the (silly/strange) CLI help order.

3. Silly for those who know more than everybody who has to read the manual or the help. The CLI help does not tell anything about setting it to ***?*** for *max supported threads*, a default or any boudaries. Setting it to 32 should fit for almost everybody as long as aomenc does not provide an auto setting.

4. Interesting  🤔   @44vince44  and me, we never had such issues while testing different sources. But I can reproduce it with your source. 🤔 

5. `rc_twopass_stats_in` is a buffer containing all of the stats packets produced in the first pass, concatenated.
So I would say it's in tool or source issue, where the stats file could only be written with one packet. I've finished another test right now with another source and it worked without problems.  🤔 
@stax76  Any suggestions?  

6. Is it possible you are using an old settings file or profile from an old StaxRip version?  With fresh settings PSNR is set to its default: 1.


Thanks for your source!

**Comment by stax76** [comment]
> Yeah, an automatic setting could be nice, indeed. At the moment it's just a default or better the first possible value.
> @stax76 Do you have the ambition to implement that?

If the input is 8-bit then defining --input-bit-depth is not strictly necessary, but if it's 10-bit then the encoder returns an error:

Fatal: Specify stream dimensions with --width (-w)  and --height (-h)

It returns that error with and without --input-bit-depth=10

I've added the custom options -w 320 -h 176, and it just hangs now, so 10-bit input don't seem to be working at all.

Maybe only change the definition to this:

.Switch = "--input-bit-depth", .Text = "Input Bit Depth", .Options = {"Automatic", "8", "10", "12"}

AlwaysOn being removed.

@Neltulz It's easier to work on tasks if every task has an independent issue/thread, for me it's too much.

**Comment by stax76** [comment]
OK, I remember for 10 bit y4m piping ffmpeg needs something like -strict -1, I'll fix that.

**Comment by Neltulz** [comment]
> If the input is 8-bit then defining --input-bit-depth is not strictly necessary, but if it's 10-bit then the encoder returns an error:

Honestly, I blame aomenc for this.  The encoder should be able to see if the video is 8, 10, 12, etc and automatically choose the correct input bit depth automatically.  This puts more work on GUI authors because you have to do extra checking and pick the best option accordingly.

IMO, Automatic input depth sounds good.  If this solution causes the correct input depth to be chosen, and causes no error with any input video, then it is the right choice.  More automation = good.

> It shouldn't really ask for w and h because it gets that info from y4m.

I have no idea why aomenc throws an error.  Have you tried my video sample? https://www.dropbox.com/s/mpgb6ht0ijz2qh3/10bit_MKV.7z?dl=0

> @Neltulz It's easier to work on tasks if every task has an independent issue/thread, for me it's too much.

Understandable.  I do not have the patience to create a separate issue for every single little thing, but I will try from now on.  I think you should spend some time test encoding various videos with aomenc before pushing a stable build, many of these issues could have been spotted and fixed if you simply performed the testing yourself, and I wouldn't have panicked and submitted 6 separate issues in this single github issue.

Also, I agree that AV1 documentation is very bad.  Like, abysmally bad.  However, that is why I spend time in the AV1 discord asking questions.  I suggest hanging around the AV1 discord, even under an alternate account so that you do not attract too much attention to yourself (if that is a major concern for you). https://discord.gg/HSBxne3

**Comment by stax76** [comment]
The --input-bit-depth issue should be fixed, I've added an automatic option and added -f yuv4mpegpipe **-strict -1**, it's a rather ugly ffmpeg quirk needing this for 10-bit.

**Comment by 44vince44** [comment]
@Neltulz Please perform your tests on the latest beta build, which means the latest Full beta withe the latest staxrip.exe available in the Dropbox.
Currently you have a staxrip.exe that has lots of improvements, and that has been released on purpose for aomenc testing.

**Comment by 44vince44** [comment]
@Neltulz 

> (...) Truthfully speaking (not to be rude), many of these issues would've been spotted immediately if you just simply test encoded a thing or 2.

There has been intense testing and improvements done by @Dendraspis and me, although we have no personal interest in that codec for now, I have been surprised that you haven't been watching and participating in the discussions, and especially the ones in the threads that you have opened. A bit of following up would be appreciated.

**Comment by Dendraspis** [comment]
> The --input-bit-depth issue should be fixed, I've added an automatic option and added -f yuv4mpegpipe **-strict -1**, it's a rather ugly ffmpeg quirk needing this for 10-bit.

@stax76 
***Automatic*** in x265 is applied through not setting the parameter and let x265 find out the right setting. For aomenc there is no documented behaviour that this works (the same way), so probably that might be problematic.  🤔 

@44vince44 
Looks like a challenge for you.  😂  In other words:  please have a look at this behaviour/setting ***if*** you plan testing aomenc  one more time .... before 3-pass mode is implemented.  😂  😜

**Comment by stax76** [comment]
@44vince44 
@Dendraspis 

I assume aomenc detects the input bit depth from y4m, and it's only necessary to set it explicitly when the input format is raw and not y4m.

**Comment by Dendraspis** [comment]
So let's hope aomenc follows those rules, too, and just lacks of full documentation. :relaxing:

**Comment by stax76** [comment]
> please provide a new staxrip.exe, if possible, and tell us when you do so, there are things that would need testing by now...

I've uploaded it.

Making now a few final tests and just upload the stable.

**Comment by stax76** [comment]
New stable version is online, hopefully without critical bug. 🤞

**Comment by Neltulz** [comment]
> I have been surprised that you haven't been watching and participating in the discussions, and especially the ones in the threads that you have opened. A bit of following up would have really been appreciated. No offense nor rudeness intended, of course. Always check you threads, they can sometimes be used for further improvements and developments.

I was in another state visiting family for 2 weeks and I did get e-mails for much of the discussion. I was loosely following it, but didn't respond because doing so on a phone would've been silly.

As for why I haven't responded since then, I stopped receiving e-mails for those issues and I assumed they were all taken care of.  I just haven't had a chance to fetch the latest StaxRip build and check out the changes that were made.

I was surprised that so much progress was being made in the month of December.  Usually that's when people are least productive due to traveling, holidays, etc etc.
